lib/service/connect.go

```diff
@@ -62,8 +64,101 @@ import (
 // reconnectToAuthService continuously attempts to reconnect to the auth
 // service until succeeds or process gets shut down
 func (process *TeleportProcess) reconnectToAuthService(role types.SystemRole) (*Connector, error) {
-	// TODO(fspmarshall): we should probably have a longer retry period for Instance certs
-	// in order to avoid catastrophic load in the event of an auth server downgrade.
+
+	// for standard service system roles, we want to check in advance if the system role is permitted
+	// by this instance's identity/join token. doing this before attempting allocation helps us avoid
+	// the situation where a user ends up creating a teleport instance that is running multiple services
+	// authorized by different join tokens, which poses challenges currently and may be a strictly
+	// invalid state in the future if we start enforcing access-control limitations on agents via
+	// their join tokens.
+	if role != types.RoleInstance && role != types.RoleAdmin && role != types.RoleAuth && process.getLocalAuth() == nil && !process.hostedPluginRoleExpected(role) {
```

Comment:
> role != types.RoleAuth && process.getLocalAuth() == nil

can we simplify this part of the predicate to just check `process.getLocalAuth() == nil` ? I think that `RoleAuth` will be amongst the roles iff `process.getLocalAuth() != nil`.